### ISSUE TITLE
Buttons that are links don't have a focus style

### ISSUE THREAD

**Issue by afercia** [WordPress/gutenberg#13267]
Broken in https://github.com/WordPress/gutenberg/pull/10031

Seems only Chrome understands `:focus:enabled` on a link. Firefox, Safari, Edge, and IE 11, don't.

I tend to think the Chrome behavior is wrong, as the `:enabled` pseudo class is supposed to work for elements [that have a corresponding disabled state](https://www.w3.org/TR/selectors-3/#enableddisabled). Links don't have a disabled state.

Currently, links styled as buttons don't have a focus style and also the style for Windows high-contrast mode doesn't work any longer.

To reproduce:
- use a keyboard
- tab to the "Preview" button (it's a link)
- in Chrome, there's a subtle focus style
- in Firefox, Safari, Edge, IE11 there's no focus style

One more example to test:
- the "View Post" button (it's a link) at the end of the Publish flow

Remove `:enabled` from this line:

https://github.com/WordPress/gutenberg/blob/36b17584439e7747a6c7800d18c94fe605b85e50/packages/components/src/button/style.scss#L174

Then build and test again: the focus style is now visible in all browsers.

Of course, this is just for testing purposes. The CSS would need some additional adjustments. /Cc @jasmussen 

Also, the current focus style is definitely too light. Focus indication needs to be _highly visible_ for obvious reasons. I'd like to discuss this with the accessibility and design team and finally get to a positive action plan 🙂

**Comment by jasmussen** [comment]
I can take a stab at both the focus style and fixing this regression. 

Short of reverting to the past behavior, do you have any instincts as to how we might address the same purpose as #10031 tried to address, but still work in all browsers? I'm sure I could think of something, but would appreciate ideas to test, so as to make this more rapidly fixed.

**Comment by afercia** [comment]
Looking at the current code, links can't be disabled because when they're, then they're rendered as buttons:

https://github.com/WordPress/gutenberg/blob/8ff1015db4bfe4343b475a65c21265fbc558afde/packages/components/src/button/index.js#L42

For example, this is what happens with the Preview button:

https://github.com/WordPress/gutenberg/blob/8ff1015db4bfe4343b475a65c21265fbc558afde/packages/editor/src/components/post-preview-button/index.js#L179-L186

When the post is not "saveable", the Preview button is not an `<a>` element, instead it's a `<button>` disabled element, with the disabled style.

So there's the need to apply the focus style specifically to links, meaning `a.is-button`. Hm to avoid a large refactoring... maybe using a class `.enabled` based on the `disabled` prop instead of `:enabled`?

**Comment by jasmussen** [comment]
Quick sidenote, I just learned that Jacobo is on paternity leave 💓— so we'll have to do without his input for the time being.

I took a look at this today, and it seems like the primary purpose of #10031 was to primarily simplify the specificity of the buttons so the !importants could be removed. That's noble on its own, but looking at the conversations it looks like disabled states played an important role here as well.

So I took a stab in https://codepen.io/joen/pen/LMqozY?editors=1100 to explore what might work. Correct, we can't use `:enabled` to increase the specificity in any browser but Chrome. But _potentially_ we could use the slightly hacky duplication trick to increase specificity enough to ditch the !importants, by simply using `:focus:focus`. That seems to work so far, I'm going to take a stab in a PR in a bit.

But the other aspect was to _not show a focus style for a disabled button_. This bends my mind a little bit now that I'm focusing on it, pardon the pun — should a disabled element ever be focusable in the first place? @afercia touches on this briefly in https://github.com/WordPress/gutenberg/pull/10031#issuecomment-424781766, but some of the details escape me, so I'll walk through the thought process:

- When a `button` is disabled, it's unfocusable, so it doesn't need special CSS to _not show the focusring_ when focused. Because it won't ever _be_ focused.
- A link (`a`) can't normally be disabled at all — we're sort of into hack territory with this, by adding an `aria-disabled="true"` property to inform screenreaders what's going on here. 

But why are we using the `a` element at all, for something that needs to able to be disabled? As far as I can tell, we're using a link for the block mover (up/down) when at the start or end of the block list. And we're using a link for the Preview button, and this can be disabled when there is no content to preview.

Presumably the Preview button is a link (`a`) because it opens a URL, but I'm not entirely sure why the block mover can't be a `button` (though it's possibly related to [this commit](https://github.com/WordPress/gutenberg/commit/6b7b1637090a05d8982c179790ecb7b6b9ac72b7)).

But in both those cases, these disabled links can still be _focused_ using the keyboard. Should they not be unfocusable when disabled? I.e. have a `tabindex="-1"`? 

If yes, then that could help unlock simpler CSS here, because we wouldn't have to provide specific CSS to "unstyle" a disabled and focused link, because a disabled link would never _be_ focused.

**Comment by afercia** [comment]
> But why are we using the a element at all, for something that needs to able to be disabled? As far as I can tell, we're using a link for the block mover (up/down) when at the start or end of the block list. And we're using a link for the Preview button, and this can be disabled when there is no content to preview.

To clarify:
- the Button component renders a button or a link depending on its props
- when there's a disabled prop, it's rendered as a button
- for example the Preview is a link, but when it has a disabled prop it becomes a button
- the Preview link doesn't have a focus style any longer after #10031
- the Preview (disabled) button shouldn't have a focus style because it's disabled > not focusable
- as far as I can tell, the block movers (up/down) are always buttons

**Comment by jasmussen** [comment]
Understood. An problem with putting `tabindex="-1"` on any link-button that's disabled?

**Comment by afercia** [comment]
We can't disable links or make the not focusable 🙂Yes it's a problem. Why we would want to do that?

**Comment by jasmussen** [comment]
Because then we wouldn't have to have a rule such as this:

```
&:focus:not(:disabled):not([aria-disabled="true"]) {
```

And also — why would we _want_ a disabled link to be focusable?

**Comment by afercia** [comment]
I think there are some big misunderstandings going on here 🙂 

For no reason, ever, a link element `<a>` should have a `disabled` or `aria-disabled` HTML attribute. Links are meant to trigger navigation and have a `href` attribute pointing to a resource URI, they can't be disabled in any way.

To target it with CSS:
> ... there's the need to apply the focus style specifically to links, meaning `a.is-button`. Hm to avoid a large refactoring... maybe using a class .enabled based on the disabled prop instead of :enabled?

**Comment by afercia** [comment]
Screenshots:

"Preview" link-button focused in Firefox (focus style not visible):

<img width="358" alt="Screenshot 2019-05-05 at 16 18 59" src="https://user-images.githubusercontent.com/1682452/57195374-ebc37480-6f51-11e9-8e73-31a188563577.png">

Works on buttons that are buttons (although it's very subtle, see #15277, #15279, and #15432)

<img width="372" alt="Screenshot 2019-05-05 at 16 19 04" src="https://user-images.githubusercontent.com/1682452/57195375-ebc37480-6f51-11e9-93ca-d6be4a03f298.png">

"View Post" link-button in the post-publish panel focused in Firefox (focus style not visible):

<img width="279" alt="Screenshot 2019-05-05 at 16 19 21" src="https://user-images.githubusercontent.com/1682452/57195422-7c01b980-6f52-11e9-8804-9250a795ccd3.png">

Works on the adjacent button:

<img width="279" alt="Screenshot 2019-05-05 at 16 19 23" src="https://user-images.githubusercontent.com/1682452/57195423-7c01b980-6f52-11e9-801c-ccc4cd272465.png">

**Comment by afercia** [comment]
Moving to the Accessibility Audit project, as #15544 greatly improved the focus style for link-buttons but those improvements are only visible in Chrome.

**Comment by aduth** [comment]
This was discussed today in the #core-editor triage session ([link requires registration](https://make.wordpress.org/chat/)):

https://wordpress.slack.com/archives/C02QB2JS7/p1557752425094500

Based on the reading that there are styles which are meant to be applied only for focus on an enabled button, and conflict with the fact that “enabled” doesn’t mean anything for links (at least in Chrome), the following action item was proposed:

Change the following line: https://github.com/WordPress/gutenberg/blob/36b17584439e7747a6c7800d18c94fe605b85e50/packages/components/src/button/style.scss#L174

To:

```scss
&:focus:not( :disabled ) {
```

...as this would apply both to (a) enabled buttons and (b) all links.

Noting that this contradicts the proposed suggestion at #11677 to avoid `:not`.

Alternatively, it could be expressed as separate style selectors, to cover both the enabled buttons and all links:

```scss
&:focus:enabled,
a&:focus,
```

**Comment by nicolad** [comment]
@aduth, updated.
Personally, I do like alternative more because it's easier to reason about without logic inversion.
But it throws me an error :(
```
Error: Invalid CSS after "a": expected "{", was "&"
"&" may only be used at the beginning of a compound selector.
```

**Comment by aduth** [comment]
@nicolad There's apparently a workaround, but it appears quite convoluted:

https://stackoverflow.com/a/25655130
